PLATFORM-2520 Log User::comparePasswords calls on production

themech opened the pull request.

themech committed:
PLATFORM-2520 log User::comparePasswords calls.

WikiaTech commented:
SonarQube analysis reported 134 issues
* ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 4 critical
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 51 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 42 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 37 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L961](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L961): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2194](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L2194): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2198](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L2198): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L3230](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L3230): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php): File "User.php" has 4892 lines, which is greater than 1000 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS104)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php#L70](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L70): Class "User" has 224 methods, which is greater than 20 authorized. Split it into smaller classes. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1448)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php#L125](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L125): Remove this commented out code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS125)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php#L131](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L131): Remove this commented out code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS125)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php#L136](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L136): Remove this commented out code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS125)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [User.php#L143](https://github.com/Wikia/app/blob/ccd6137fa6a36ffb6cccb99673baae42f8408e2a/includes/User.php#L143): Remove this commented out code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS125)

WikiaTech commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
 	public static function comparePasswords( $hash, $password, $userId = false ) {
+		// Wikia change - begin
+		// @see PLATFORM-2502 comparing new passwords in PHP code.
+		// @todo mech remove after the new password hashing is implemented.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this "TODO" comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)

themech commented:
@harnash please take a look.

harnash commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
 	public static function comparePasswords( $hash, $password, $userId = false ) {
+		// Wikia change - begin
+		// @see PLATFORM-2502 comparing new passwords in PHP code.
+		// @todo mech remove after the new password hashing is implemented.
```
I would suggest create follow up ticket to cleanup the code so we won't loose it when this gets reassigned to someone else

themech replied:
Good idea, I've created PLATFORM-2526

Grunny commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
+			'NEW_HASHING comparePasswords called in PHP',
+			[
+				'user_id' => $userId,
+				'caller' => $caller['class'].'::'.$caller['function'],
```
`wfGetCaller()` should do this for you.

themech replied:
As mentioned in the meantime, wfGetCaller was broken. I'm including a MW fix (supplied by @Grunny) as a part of this PR and switching my code to use it.

harnash commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
+		// @see PLATFORM-2502 comparing new passwords in PHP code.
+		// @todo mech remove after the new password hashing is implemented.
+		$caller = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS)[1];
+		Wikia\Logger\WikiaLogger::instance()->debug(
```
Shouldn't we sample this first to see what scale we are dealing with? We can then remove/increase sampling rate if we are sure we are not killing our ELK?

themech replied:
I believe Helios handles most of the password checking requests and logs the calls without any sampling (https://github.com/Wikia/app/blob/b03df0a89ed672697e9c130d529bf1eb25f49cda/extensions/wikia/Helios/User.class.php#L161). Additionally I've checked those logs on the dashboard (https://kibana.wikia-inc.com/index.html#/dashboard/elasticsearch/HELIOS_LOGIN) and it generates round 0.1 qps

harnash replied:
Alright. Let us be weary if we notice some unexpected spike in generated logs after this goes live.

harnash commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
+			[
+				'user_id' => $userId,
+				'caller' => $caller['class'].'::'.$caller['function'],
+				'exception' => new Exception()
```
Why exception here? Empty?

themech replied:
To log the stacktrace, so it will be easier to see where the call came from and fix it to use Helios. I had an option of using Exception handling already supported by WikiaLogger or formatting the result of debug_backtrace myself. Also, I think we use this approach in couple of other places.

harnash replied:
OK. Got it.

Grunny committed:
Add fix for wfGetCaller from newer versions of MW

themech committed:
PR comments - created a TODO-related ticket and used wfGetCaller

WikiaTech commented on includes/User.php:
```diff
@@ -4218,6 +4218,20 @@ public static function crypt( $password, $salt = false ) {
 	public static function comparePasswords( $hash, $password, $userId = false ) {
+		// Wikia change - begin
+		// @see PLATFORM-2502 comparing new passwords in PHP code.
+		// @todo mech remove after the new password hashing is implemented (PLATFORM-2526).
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this "TODO" comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1135)

WikiaTech commented:
SonarQube analysis reported 224 issues
* ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 4 critical
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 91 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 66 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 63 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L961](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/User.php#L961): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2194](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/User.php#L2194): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2198](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/User.php#L2198): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L3230](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/User.php#L3230): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php): File "GlobalFunctions.php" has 4098 lines, which is greater than 1000 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS104)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L66](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php#L66): Remove the unused function parameter "$var". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L70](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php#L70): Remove the unused function parameters "$var, $level". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L72](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php#L72): Remove the unused function parameters "$var, $level". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L140](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php#L140): Reduce the number of returns of this function 4, down to the maximum allowed 3. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1142)
1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L196](https://github.com/Wikia/app/blob/2aee0a6ccaf63188858f44c23d176a7d6d291698/includes/GlobalFunctions.php#L196): Remove the unused function parameter "$array1". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

themech committed:
Removing obsolete comment.

WikiaTech commented:
SonarQube analysis reported 224 issues
- ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) 4 critical
- ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 91 major
- ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 66 minor
- ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 63 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L961](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/User.php#L961): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
2. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2194](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/User.php#L2194): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
3. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L2198](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/User.php#L2198): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
4. ![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) [User.php#L3230](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/User.php#L3230): Remove this hard-coded password. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2068)
5. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php): File "GlobalFunctions.php" has 4098 lines, which is greater than 1000 authorized. Split it into smaller files. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS104)
6. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L66](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php#L66): Remove the unused function parameter "$var". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
7. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L70](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php#L70): Remove the unused function parameters "$var, $level". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
8. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L72](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php#L72): Remove the unused function parameters "$var, $level". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)
9. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L140](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php#L140): Reduce the number of returns of this function 4, down to the maximum allowed 3. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1142)
10. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [GlobalFunctions.php#L196](https://github.com/Wikia/app/blob/622f98cebd8b146d13267c869bd5cea390c89bab/includes/GlobalFunctions.php#L196): Remove the unused function parameter "$array1". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1172)

themech commented:
Addressed all the comments, also including wfGetCaller fix in this PR (ported by @Grunny from https://github.com/wikimedia/mediawiki/commit/2e506c4809624c2a91e7798a38f2996e321cd95f). I will of course keep my eye on the new message in Kibana to make sure we're not overloading the ELK.

@harnash, @Grunny please take another look.

Grunny reviewed (approved):
LGTM.

ludwikkazmierczak commented:
Tested and Accepted
